src/gc.c

```diff
@@ -2125,6 +2292,9 @@ mark: {
             int stkbuf = (ta->stkbuf != (void*)(intptr_t)-1 && ta->stkbuf != NULL);
             int16_t tid = ta->tid;
             jl_ptls_t ptls2 = jl_all_tls_states[tid];
+            ptls->last_gc_mark_sp = &sp;
```

Comment:
I think I mentioned this in an early comment. This leaks the address of `sp`, should be replaced by copy instead.

Reply:
I don't understand what you mean by "replaced by copy" here?

And I don't think you mentioned this before (nor could I find such a comment). To the contrary, our first version of the Julep and the code in our branch (from before the PR was submitted) did not have `last_gc_mark_sp` but rather used a more complicated approach; it was then [suggested to us by *you*](https://github.com/JuliaLang/Juleps/pull/50#discussion_r198834823) to move this into the TLS (for which we were quite grateful, as it simplified some things considerably).


Anyway, perhaps you could elaborate on what you imagine should replace this (while ensuring that `jl_gc_mark_queue_obj` and `jl_gc_mark_queue_objarray` keep working)

Reply:
> so it's better to just explicitly synchronize them at those point instead of leaking `&sp` from the beginning of the function.

That's actually what i mean by explicitly synchronize instead of leaking address. So instead of

```c++
ptls->last_gc_mark_sp = &sp;
<external call>
```

Do

```c++
ptls->last_gc_mark_sp = sp;
<external call>
sp = ptls->last_gc_mark_sp;
```

Reply:
OK, sure, we can do that.

However (with the caveat that I have not yet had a chance to try the change), I think that means either including gc.h from julia_threads.h to get `gc_mark_sp_t`, or else using a hard coded buffer size there matching the size of `gc_mark_sp_t` (and then having a `static_assert` somewhere to ensure that value stays correct). Which do you prefer? Resp. if you have an alternate idea, please let us know?

Reply:
def of `gc_mark_sp_t` could be moved to `julia_thread.h` and renamed to `jl_gc_mark_sp_t`. That's the approach used by `jl_gc_mark_cache_t`.

Reply:
About copying: this could introduce measurable overhead, as the struct consists of eight words. More importantly, as the contents of the struct change during marking, this would be problematic during root scanning, as there would be no way to retrieve any changes to the struct. And the internals of the struct would be exposed, anyway.

An alternative idea would be to explicitly document when the field is valid.

Renaming the type wouldn't be a problem, of course. Part of the reason why I made the field a `void *` was for purposes of information hiding, i.e. to make it opaque and to not expose its internals. But that could also be done with an opaque struct pointer.

Reply:
@rbehrends well, can you measure it? If you implement this in a commit atop the PR, we can easily drop it later on if it turns out to be too bad.

I don't understand what problem with "contents of the struct change during marking" is; and what changes you think need to be "retrieved". Could you elaborate?

As to you "An alternative idea", I also don't quite see to what it is meant to be an alternative?

Reply:
> About copying: this could introduce measurable overhead, as the struct consists of eight words.

The issue is that the control flow in this function is very unfriendly to the compiler, to put it mildly, so the compiler won't be able to do any escape analysis following the control flow. The `sp` is the most accessed variable in the function so I do not want to compromise it's performance. (Note that all the fast path in this function is fully inlined and `sp` doesn't leak to any other functions). Also note that since there aren't that many callee saved registers on x86, all of the function calls are already surrounded by many register spills which are basically copying things around.

> as there would be no way to retrieve any changes to the struct

That's why they are copied back?

Reply:
Hmm, I see the problem.

> That's why they are copied back?

Hmm, that would be two struct copies per mark operation. I really have to benchmark that to see what happens. It may be necessary to add to the API to permit bulk operations. I'll go and do a provisional implementation of this approach and see how that turns out, performance-wise.

Reply:
Some initial benchmarking seems to indicate that the overhead remains fairly small, even if you have lots of objects with custom mark functions. I still need to figure out why I now have a rare and difficult to reproduce crash as the result of these changes, though.